src/TwigEnvironmentFactory.php

```diff
@@ -0,0 +1,184 @@
+            : [];
+        $this->injectExtensions($environment, $container, $extensions);
+
+        // Add template paths
+        $allPaths = isset($config['paths']) && is_array($config['paths']) ? $config['paths'] : [];
+        foreach ($allPaths as $namespace => $paths) {
+            $namespace = is_numeric($namespace) ? null : $namespace;
+            $namespace = $namespace ?: TwigFilesystem::MAIN_NAMESPACE;
+            foreach ((array) $paths as $path) {
+                $loader->addPath($path, $namespace);
+            }
+        }
+
+        // Inject environment
+        return $environment;
+    }
+
+    /**
+     * Inject extensions into the TwigEnvironment instance.
+     *
+     * @param TwigEnvironment    $environment
+     * @param ContainerInterface $container
+     * @param array              $extensions
+     *
+     * @throws Exception\InvalidExtensionException
+     */
+    private function injectExtensions(TwigEnvironment $environment, ContainerInterface $container, array $extensions)
+    {
+        foreach ($extensions as $extension) {
+            // Load the extension from the container
+            if (is_string($extension) && $container->has($extension)) {
+                $extension = $container->get($extension);
+            }
+
+            if (!$extension instanceof TwigExtensionInterface) {
+                throw new Exception\InvalidExtensionException(sprintf(
+                    'Twig extension must be an instance of Twig_ExtensionInterface; "%s" given,',
+                    is_object($extension) ? get_class($extension) : gettype($extension)
+                ));
+            }
+
+            if ($environment->hasExtension($extension->getName())) {
+                continue;
+            }
+
+            $environment->addExtension($extension);
+        }
+    }
+
+    /**
+     * Merge expressive templating config with twig config.
+     *
+     * Pulls the `templates` and `twig` top-level keys from the configuration,
+     * if present, and then returns the merged result, with those from the twig
+     * array having precedence.
+     *
+     * @param array|ArrayObject $config
+     *
+     * @return array
+     * @throws Exception\InvalidConfigException if a non-array, non-ArrayObject
+     *     $config is received.
+     */
+    private function mergeConfig($config)
+    {
+        $config = $config instanceof ArrayObject ? $config->getArrayCopy() : $config;
+
+        if (!is_array($config)) {
+            throw new Exception\InvalidConfigException(sprintf(
+                'config service MUST be an array or ArrayObject; received %s',
+                is_object($config) ? get_class($config) : gettype($config)
+            ));
+        }
+
+        $expressiveConfig = (isset($config['templates']) && is_array($config['templates']))
+            ? $config['templates']
+            : [];
+        $twigConfig       = (isset($config['twig']) && is_array($config['twig']))
```

Comment:
Since the previous assignment is several lines above, no need to align the `=` operators.